File: backend/internal/janusconfig/catalog.go
package janusconfig

import "fmt"

// Owner identifies the subsystem that must give an assignment semantics.
// Classification is deliberately separate from parsing: parsing is lossless,
// while validation must reject any assignment whose behavior is unknown.
type Owner string

const (
	OwnerTopology   Owner = "topology"
	OwnerHardware   Owner = "hardware"
	OwnerRunControl Owner = "run_control"
	OwnerStorage    Owner = "storage"
	OwnerAnalysis   Owner = "analysis"
)

// ClassifiedAssignment pairs a losslessly parsed assignment with its declared
// semantic owner.
type ClassifiedAssignment struct {
	Assignment Assignment
	Owner      Owner
}

var settingOwners = map[string]Owner{
	"Open": OwnerTopology,

	"HV_Vbias": OwnerHardware, "HV_Imax": OwnerHardware,
	"HV_Adjust_Range": OwnerHardware, "HV_IndivAdj": OwnerHardware,
	"TempSensType": OwnerHardware, "TempFeedbackCoeff": OwnerHardware,
	"EnableTempFeedback": OwnerHardware, "AcquisitionMode": OwnerHardware,
	"EnableToT": OwnerHardware, "BunchTrgSource": OwnerHardware,
	"VetoSource": OwnerHardware, "ValidationSource": OwnerHardware,
	"ValidationMode": OwnerHardware, "CountingMode": OwnerHardware,
	"ChTrg_Width": OwnerHardware, "EnableCntZeroSuppr": OwnerHardware,
	"TrgIdMode": OwnerHardware, "TriggerLogic": OwnerHardware,
	"Tlogic_Width": OwnerHardware, "MajorityLevel": OwnerHardware,
	"PtrgPeriod": OwnerHardware, "TrefSource": OwnerHardware,
	"TrefWindow": OwnerHardware, "TrefDelay": OwnerHardware,
	"T0_Out": OwnerHardware, "T1_Out": OwnerHardware,
	"ChEnableMask0": OwnerHardware, "ChEnableMask1": OwnerHardware,
	"FastShaperInput": OwnerHardware, "TD_CoarseThreshold": OwnerHardware,
	"TD_FineThreshold": OwnerHardware, "Hit_HoldOff": OwnerHardware,
	"Tlogic_Mask0": OwnerHardware, "Tlogic_Mask1": OwnerHardware,
	"QD_CoarseThreshold": OwnerHardware, "QD_FineThreshold": OwnerHardware,
	"Q_DiscrMask0": OwnerHardware, "Q_DiscrMask1": OwnerHardware,
	"GainSelect": OwnerHardware, "HG_Gain": OwnerHardware,
	"LG_Gain": OwnerHardware, "Pedestal": OwnerHardware,
	"ZS_Threshold_LG": OwnerHardware, "ZS_Threshold_HG": OwnerHardware,
	"HG_ShapingTime": OwnerHardware, "LG_ShapingTime": OwnerHardware,
	"HoldDelay": OwnerHardware, "MuxClkPeriod": OwnerHardware,
	"AnalogProbe0": OwnerHardware, "DigitalProbe0": OwnerHardware,
	"ProbeChannel0": OwnerHardware, "AnalogProbe1": OwnerHardware,
	"DigitalProbe1": OwnerHardware, "ProbeChannel1": OwnerHardware,
	"TestPulseSource": OwnerHardware, "TestPulseAmplitude": OwnerHardware,
	"TestPulseDestination": OwnerHardware, "TestPulsePreamp": OwnerHardware,

	"StartRunMode": OwnerRunControl, "StopRunMode": OwnerRunControl,
	"EventBuildingMode": OwnerRunControl, "TstampCoincWindow": OwnerRunControl,
	"PresetTime": OwnerRunControl, "PresetCounts": OwnerRunControl,
	"JobFirstRun": OwnerRunControl, "JobLastRun": OwnerRunControl,
	"RunSleep": OwnerRunControl, "EnableJobs": OwnerRunControl,
	"RunNumber_AutoIncr": OwnerRunControl,

	"DataFilePath": OwnerStorage, "OF_OutFileUnit": OwnerStorage,
	"OF_EnMaxSize": OwnerStorage, "OF_MaxSize": OwnerStorage,
	"OF_RawData": OwnerStorage, "OF_ListBin": OwnerStorage,
	"OF_ListAscii": OwnerStorage, "OF_ListCSV": OwnerStorage,
	"OF_Sync": OwnerStorage, "OF_ServiceInfo": OwnerStorage,
	"OF_RunInfo": OwnerStorage,

	"DataAnalysis": OwnerAnalysis, "EnableListZeroSuppr": OwnerAnalysis,
	"OF_SpectHisto": OwnerAnalysis, "OF_ToAHisto": OwnerAnalysis,
	"OF_ToTHisto": OwnerAnalysis, "OF_MCS": OwnerAnalysis,
	"OF_Staircase": OwnerAnalysis, "EHistoNbin": OwnerAnalysis,
	"ToAHistoNbin": OwnerAnalysis, "ToARebin": OwnerAnalysis,
	"ToAHistoMin": OwnerAnalysis, "MCSHistoNbin": OwnerAnalysis,
}

// Classify rejects the whole document if any setting lacks an explicit owner.
// This prevents syntactically valid production fields from being ignored.
func (d *Document) Classify() ([]ClassifiedAssignment, error) {
	classified := make([]ClassifiedAssignment, 0, len(d.Assignments))
	for _, assignment := range d.Assignments {
		owner, ok := settingOwners[assignment.Name]
		if !ok {
			return nil, fmt.Errorf("line %d: unsupported JANUS setting %q", assignment.Line, assignment.Name)
		}
		classified = append(classified, ClassifiedAssignment{Assignment: assignment, Owner: owner})
	}
	return classified, nil
}

File: frontend/src/janus/catalog.ts
import definitions from './param_defs_5.0.0.txt?raw'

export type JanusScope = 'global' | 'board' | 'channel'
export type JanusWidget = 'boolean' | 'select' | 'integer' | 'number' | 'text' | 'monitor'

export interface JanusDependency {
  parameter: string
  values: string[]
}

export interface JanusParameter {
  name: string
  section: string
  defaultValue: string
  scope: JanusScope
  widget: JanusWidget
  description: string
  options: string[]
  min?: number
  max?: number
  step?: number
  units?: string[]
  activeWhen?: JanusDependency
}

const constraints: Record<string, Pick<JanusParameter, 'min' | 'max' | 'step' | 'units'>> = {
  HV_Vbias: { min: 20, max: 85, step: 0.1, units: ['V', 'mV', 'uV'] },
  HV_Imax: { min: 0, step: 0.1, units: ['mA', 'uA', 'A'] },
  HV_IndivAdj: { min: 0, max: 255, step: 1 },
  TempFeedbackCoeff: { step: 0.1, units: ['mV/degC'] },
  TstampCoincWindow: { min: 0, step: 1, units: ['ns', 'us', 'ms', 's'] },
  PresetTime: { min: 0, step: 1, units: ['s', 'ms', 'us', 'ns'] },
  PresetCounts: { min: 0, step: 1 },
  JobFirstRun: { min: 0, step: 1 },
  JobLastRun: { min: 0, step: 1 },
  RunSleep: { min: 0, step: 1, units: ['s', 'ms', 'us', 'ns'] },
  OF_MaxSize: { min: 1, step: 1, units: ['MB', 'GB'] },
  ChTrg_Width: { min: 8, max: 2032, step: 8, units: ['ns', 'us'] },
  Tlogic_Width: { min: 0, step: 8, units: ['ns', 'us'] },
  MajorityLevel: { min: 1, max: 64, step: 1 },
  PtrgPeriod: { min: 0, step: 8, units: ['ns', 'us', 'ms', 's'] },
  TrefWindow: { min: 0, step: 8, units: ['ns', 'us', 'ms', 's'] },
  TrefDelay: { min: -4194304, max: 4194296, step: 1, units: ['ns', 'us', 'ms', 's'] },
  TD_CoarseThreshold: { min: 0, max: 2047, step: 1 },
  TD_FineThreshold: { min: 0, max: 15, step: 1 },
  Hit_HoldOff: { min: 0, step: 8, units: ['ns', 'us', 'ms', 's'] },
  QD_CoarseThreshold: { min: 0, max: 2047, step: 1 },
  QD_FineThreshold: { min: 0, max: 15, step: 1 },
  HG_Gain: { min: 1, max: 63, step: 1 },
  LG_Gain: { min: 1, max: 63, step: 1 },
  Pedestal: { min: 0, max: 16383, step: 1 },
  ZS_Threshold_LG: { min: 0, max: 65535, step: 1 },
  ZS_Threshold_HG: { min: 0, max: 65535, step: 1 },
  HoldDelay: { min: 0, step: 1, units: ['ns', 'us'] },
  MuxClkPeriod: { min: 0, step: 1, units: ['ns', 'us'] },
  ToARebin: { min: 1, step: 1 },
  ToAHistoMin: { step: 1, units: ['ns', 'us', 'ms', 's'] },
  ProbeChannel0: { min: 0, max: 31, step: 1 },
  ProbeChannel1: { min: 32, max: 63, step: 1 },
  TestPulseAmplitude: { min: 0, max: 4095, step: 1 },
}

const dependencies: Record<string, JanusDependency> = {
  ExtRunSource: { parameter: 'StartRunMode', values: ['TDL_EXTRUN'] },
  GPSTimeUTC: { parameter: 'StartRunMode', values: ['TDL_GPS'] },
  TstampCoincWindow: {
    parameter: 'EventBuildingMode',
    values: ['TRGTIME_SORTING', 'TRGID_SORTING'],
  },
  PresetTime: { parameter: 'StopRunMode', values: ['PRESET_TIME'] },
  PresetCounts: { parameter: 'StopRunMode', values: ['PRESET_COUNTS'] },
  JobFirstRun: { parameter: 'EnableJobs', values: ['1'] },
  JobLastRun: { parameter: 'EnableJobs', values: ['1'] },
  RunSleep: { parameter: 'EnableJobs', values: ['1'] },
  OF_MaxSize: { parameter: 'OF_EnMaxSize', values: ['1'] },
  ChTrg_Width: { parameter: 'CountingMode', values: ['PAIRED_AND'] },
  MajorityLevel: { parameter: 'TriggerLogic', values: ['MAJ64', 'MAJ32_AND2'] },
  TestPulseAmplitude: {
    parameter: 'TestPulseSource',
    values: ['EXT', 'T0-IN', 'T1-IN', 'PTRG', 'SW-CMD'],
  },
  TestPulseDestination: {
    parameter: 'TestPulseSource',
    values: ['EXT', 'T0-IN', 'T1-IN', 'PTRG', 'SW-CMD'],
  },
  TestPulsePreamp: {
    parameter: 'TestPulseSource',
    values: ['EXT', 'T0-IN', 'T1-IN', 'PTRG', 'SW-CMD'],
  },
}

function unquote(value: string) {
  return value.startsWith('"') && value.endsWith('"') ? value.slice(1, -1) : value
}

function widget(type: string): JanusWidget {
  if (type === 'm') return 'monitor'
  if (type === 'b') return 'boolean'
  if (type === 'c') return 'select'
  if (type.startsWith('d')) return 'integer'
  if (type.startsWith('f') || type === 'u') return 'number'
  return 'text'
}

function parseDefinitions(source: string): JanusParameter[] {
  let section = ''
  let current: JanusParameter | undefined
  const result: JanusParameter[] = []
  for (const line of source.split(/\r?\n/)) {
    const heading = line.match(/^\[([^\]]+)\]$/)
    if (heading) {
      section = heading[1]
      continue
    }
    const option = line.match(/^\s+-\s+(.+?)\s*$/)
    if (option && current) {
      current.options.push(unquote(option[1]))
      continue
    }
    const match = line.match(
      /^([A-Za-z][A-Za-z0-9_]*)\s+("[^"]*"|\S+)\s+([gbc-])\s+([a-z-]+)\s*(?:#\s*(.*))?$/,
    )
    if (!match || match[3] === '-' || match[4] === '-') {
      current = undefined
      continue
    }
    current = {
      name: match[1],
      section,
      defaultValue: unquote(match[2]),
      scope: ({ g: 'global', b: 'board', c: 'channel' } as const)[match[3] as 'g' | 'b' | 'c'],
      widget: widget(match[4]),
      description: match[5]?.trim() ?? '',
      options: [],
      ...constraints[match[1]],
      activeWhen: dependencies[match[1]],
    }
    result.push(current)
  }
  return result
}

export const janusParameters = parseDefinitions(definitions)
export const janusParameterCatalog = new Map(
  janusParameters.map((parameter) => [parameter.name, parameter]),
)
